File: backend/src/main/java/com/dativa/imports/ImportFingerprint.java
package com.dativa.imports;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.Normalizer;
import java.time.Instant;
import java.util.Locale;

final class ImportFingerprint {
    private ImportFingerprint() {}

    static String of(
            String dataset,
            Instant occurredAt,
            String region,
            String category,
            String product,
            String seller,
            int quantity,
            BigDecimal unitPrice,
            BigDecimal amount) {
        return String.join(
                "|",
                dataset == null ? "" : dataset,
                occurredAt == null ? "" : Long.toString(occurredAt.getEpochSecond()),
                normalize(region),
                normalize(category),
                normalize(product),
                normalize(seller),
                Integer.toString(quantity),
                money(unitPrice),
                money(amount));
    }

    static String normalize(String value) {
        if (value == null) {
            return "";
        }
        String nfd = Normalizer.normalize(value.trim(), Normalizer.Form.NFD);
        return nfd.replaceAll("\\p{M}+", "").toLowerCase(Locale.ROOT).replaceAll("\\s+", " ");
    }

    private static String money(BigDecimal value) {
        BigDecimal amount = value == null ? BigDecimal.ZERO : value;
        return amount.setScale(2, RoundingMode.HALF_UP).toPlainString();
    }
}

File: frontend/apps/dativa-web/src/app/features/import/data/import-fingerprint.ts
import { DatasetId } from '../../dashboards/widgets/widget.models';
import { TransactionRow } from '../../explorer/data/transaction.models';

export type ImportDuplicateKind = 'file' | 'existing';

export function rowFingerprint(row: {
  dataset: DatasetId | string;
  occurredAt: string;
  region: string;
  category: string;
  product: string;
  seller: string;
  quantity: number;
  unitPrice: number;
  amount: number;
}): string {
  return [
    row.dataset,
    fingerprintTime(row.occurredAt),
    normalizeImportText(row.region),
    normalizeImportText(row.category),
    normalizeImportText(row.product),
    normalizeImportText(row.seller),
    String(row.quantity),
    money(row.unitPrice),
    money(row.amount),
  ].join('|');
}

export function normalizeImportText(value: string): string {
  return value
    .normalize('NFD')
    .replace(/\p{Diacritic}/gu, '')
    .trim()
    .toLowerCase()
    .replace(/\s+/g, ' ');
}

export function fingerprintSet(rows: readonly TransactionRow[], dataset?: DatasetId): Set<string> {
  const selected = dataset ? rows.filter((row) => row.dataset === dataset) : rows;
  return new Set(selected.map((row) => rowFingerprint(row)));
}

function fingerprintTime(iso: string): string {
  const time = Date.parse(iso);
  if (!Number.isFinite(time)) {
    return iso;
  }
  return String(Math.floor(time / 1000));
}

function money(value: number): string {
  return (Math.round(value * 100) / 100).toFixed(2);
}
